[clang] Adjust -mlarge-data-threshold handling

aeubanks opened the pull request:
Make it apply to x86-64 medium and large code models since that's what the backend does.

Limit logic to exclude x86-32.

Default to 0, let the driver set it to 65536 for the medium code model
if one is not passed. Set it to 0 for the large code model by default to
match gcc and since some users make assumptions about the large code
model that any small data will break.

aeubanks committed:
[clang] Adjust -mlarge-data-threshold handling

Make it apply to x86-64 medium and large code models since that's what the backend does.

Warn if it's used for x86-32.

Default to 0, let the driver set it to 65536 for the medium code model
if one is not passed. Set it to 0 for the large code model by default to
match gcc and since some users make assumptions about the large code
model that any small data will break.

MaskRay commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5752,20 +5752,29 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
+      CmdArgs.push_back("-mlarge-data-threshold=0");
+    }
+  } else {
+    if (Arg *A = Args.getLastArg(options::OPT_mlarge_data_threshold_EQ)) {
```
You can remove `err_drv_unsupported_opt_for_target` here in favor of `TargetSpecific` at `def mlarge_data_threshold_EQ`

aeubanks committed:
fixups

`Flags<[TargetSpecific]>`
remove x86-32 test since the target doesn't accept mcmodel=medium anyway
